### PR TITLE
Improve some documentation

### LINKED ISSUE #402: Parser::map_with_state example actually shows Parser::map_with_span
The doc comment for map_with_state looks like this (note that the example doesn't show map_with_state being used):

    /// Map the output of this parser to another value, making use of the parser's state when doing so.
    ///
    /// This is very useful for parsing non context-free grammars.
    ///
    /// The output type of this parser is `U`, the same as the function's output.
    ///
    /// # Examples
    ///
    /// ```
    /// # use chumsky::prelude::*;
    /// use std::ops::Range;
    ///
    /// // It's common for AST nodes to use a wrapper type that allows attaching span information to them
    /// #[derive(Debug, PartialEq)]
    /// pub struct Spanned<T>(T, SimpleSpan<usize>);
    ///
    /// let ident = text::ident::<_, _, extra::Err<Simple<char>>>()
    ///     .map_with_span(|ident, span| Spanned(ident, span))
    ///     .padded();
    ///
    /// assert_eq!(ident.parse("hello").into_result(), Ok(Spanned("hello", (0..5).into())));
    /// assert_eq!(ident.parse("       hello   ").into_result(), Ok(Spanned("hello", (7..12).into())));
    /// ```
    fn map_with_state<U, F: Fn(O, I::Span, &mut E::State) -> U>(
        self,
        f: F,
    ) -> MapWithState<Self, O, F>
    where
        Self: Sized,
    {
        MapWithState {
            parser: self,
            mapper: f,
            phantom: EmptyPhantom::new(),
        }
    }

### PR DESCRIPTION
**Pull request by CraftSpider**
Fixes #402
Steps towards #403, but not a full fix

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -524,16 +535,61 @@ pub trait Parser<'a, I: Input<'a>, O, E: ParserExtra<'a, I> = extra::Default>:
...
+    /// #                     self.strings.push(str.to_string());
+    /// #                     Key(self.strings.len() - 1)
+    /// #                 }
+    /// #             }
+    /// #         }
+    /// #         pub fn get(&self, key: Key) -> &str {
+    /// #             &self.strings[key.0]
+    /// #         }
+    /// #     }
+    /// # }
```
Comment (zesterer) [inline src/lib.rs:560]: Good example! I'd even suddenly making the interner itself visible, or spending on `lasso`

File: src/lib.rs [hunk 2]
```diff
@@ -806,7 +862,15 @@ pub trait Parser<'a, I: Input<'a>, O, E: ParserExtra<'a, I> = extra::Default>:
     }
 
     /// Parse input as part of a token-tree - using an input generated from within the current
-    /// input.
+    /// input. In other words, this parser will attempt to create a *new* input stream from within
+    /// the one it is being run on, and the parser is was called on will be provided this *new* input.
```
Comment (zesterer) [inline src/lib.rs:866]: is -> it

File: src/lib.rs (hunk 1 above)
Comment (CraftSpider) [reply src/lib.rs:560]: By 'spending' do you mean pulling it in and actually using it? If you're fine with adding it as a dev dependency, I think that would be good - I'd be hesitant to expose the stub interner here - it's not the *worst* implementation but it's not one I'd advise exactly.

Comment (zesterer) [reply src/lib.rs:560]: Sorry, I think I'd had a bit to drink and was typing on mobile when I wrote this. Should have been:

> Good example! I'd even suggest making the interner itself visible or depending on `lasso` (i.e: using `lasso` is part of the doc example given that it's a commonly used interning crate)

**Comment by zesterer** [comment]
Happy to merge?

**Comment by CraftSpider** [comment]
Yep, ready when you are

**Pull request merged**